python_modules/dagster-wandb/dagster_wandb/io_manager.py

```diff
@@ -0,0 +1,565 @@
+import datetime
+import os
+import pickle
+import platform
+import shutil
+import sys
+import time
+from contextlib import contextmanager
+from typing import List, Optional
+
+from wandb.sdk.data_types.base_types.wb_value import WBValue
+from wandb.sdk.wandb_artifacts import Artifact
+
+from dagster import (
+    IOManager,
+    InitResourceContext,
+    InputContext,
+    MetadataValue,
+    OutputContext,
+    io_manager,
+)
+
+from .resources import WANDB_CLOUD_HOST
+from .version import __version__
+
+if sys.version_info >= (3, 8):
+    from typing import TypedDict
+else:
+    from typing_extensions import TypedDict
+
+try:
+    import dill
+
+    has_dill = True
+except ImportError:
+    has_dill = False
+
+try:
+    import cloudpickle
+
+    has_cloudpickle = True
+except ImportError:
+    has_cloudpickle = False
+
+try:
+    import joblib
+
+    has_joblib = True
+except ImportError:
+    has_joblib = False
+
+
+PICKLE_FILENAME = "output.pickle"
+DILL_FILENAME = "output.dill"
+CLOUDPICKLE_FILENAME = "output.cloudpickle"
+JOBLIB_FILENAME = "output.joblib"
+ACCEPTED_SERIALIZATION_MODULES = [
+    "dill",
+    "cloudpickle",
+    "joblib",
+    "pickle",
+]
+
+
+class Config(TypedDict):
+    dagster_run_id: str
+    wandb_host: str
+    wandb_entity: str
+    wandb_project: str
+    wandb_run_name: Optional[str]
+    wandb_run_id: Optional[str]
+    wandb_run_tags: Optional[List[str]]
+    cache_duration_in_minutes: Optional[int]
+
+
+class WandbArtifactsIOManagerError(Exception):
+    def __init__(self, message="A W&B Artifacts IO Manager error occurred."):
+        self.message = message
+        super().__init__(self.message)
+
+
+class ArtifactsIOManager(IOManager):
+    """IO Manager to handle Artifacts in Weights & Biases (W&B) .
+
+    It handles 3 different inputs:
+    - Pickable objects (the serialization module is configurable)
+    - W&B Objects (Audio, Table, Image, etc)
+    - W&B Artifacts
+    """
+
+    def __init__(self, wandb_client, config: Config):
```

Comment:
we could add a `base_dir` arg here to allow users to specify their own `base_dir`; if it's not provided by the user, we can default to the dagster system base dir. see in [1]